src/DropdownMenu/DropdownMenu.tsx

```diff
@@ -0,0 +1,158 @@
+   * A callback which receives the selected item or `undefined` when an item is activated in the menu.  If the activated item is the same as the current
+   * `selectedItem`, `undefined` will be passed.
+   */
+  setSelectedItem?: (item?: ItemInput) => unknown
+}
+
+/**
+ * A `DropdownMenu` provides an anchor (button by default) that will open a floating menu of selectable items.  The menu can be
+ * opened and navigated using keyboard or mouse.  When an item is selected, the menu will close and the anchor contents will be updated
+ * with the selection.
+ */
+export function DropdownMenu({
+  renderAnchor = <T extends DropdownButtonProps>(props: T) => <DropdownButton {...props} />,
+  renderItem = Item,
+  placeholder,
+  selectedItem,
+  setSelectedItem,
+  ...listProps
+}: DropdownMenuProps): JSX.Element {
+  const anchorRef = useRef<HTMLElement>(null)
+  const [overlayRef, updateOverlayRef] = useRenderForcingRef<HTMLDivElement>()
+
+  const anchorId = `dropdownMenuAnchor-${randomId()}`
+
+  const [open, setOpen] = useState<boolean>(false)
+  const [focusType, setFocusType] = useState<null | 'anchor' | 'list'>(null)
+  const onDismiss = useCallback(() => {
+    setOpen(false)
+    setFocusType(null)
+  }, [])
+
+  const onAnchorKeyDown = useCallback(
+    (event: React.KeyboardEvent<HTMLElement>) => {
+      if (!event.defaultPrevented) {
+        if (!open) {
+          if (event.key === 'ArrowDown' || event.key === 'ArrowUp') {
+            setFocusType('list')
+            setOpen(true)
+            event.preventDefault()
+          } else if (event.key === ' ' || event.key === 'Enter') {
+            setFocusType('anchor')
+            setOpen(true)
+            event.preventDefault()
+          }
+        } else if (focusType === 'anchor') {
+          if (['ArrowDown', 'ArrowUp', 'Tab', 'Enter'].indexOf(event.key) !== -1) {
+            setFocusType('list')
+            event.preventDefault()
+          } else if (event.key === 'Escape') {
+            onDismiss()
+            event.preventDefault()
+          }
+        }
+      }
+    },
+    [open, focusType, onDismiss]
+  )
+  const onAnchorClick = useCallback(
+    (event: React.MouseEvent<HTMLElement>) => {
+      if (!event.defaultPrevented && event.button === 0 && !open) {
+        setOpen(true)
+        setFocusType('anchor')
+      }
+    },
+    [open]
+  )
+
+  const {position} = useAnchoredPosition({anchorElementRef: anchorRef, floatingElementRef: overlayRef})
+
+  useFocusZone({containerRef: overlayRef, disabled: !open || focusType !== 'list' || !position})
+  useFocusTrap({containerRef: overlayRef, disabled: !open || focusType !== 'list' || !position})
+
+  return (
+    <>
+      {renderAnchor({
+        ref: anchorRef,
+        id: anchorId,
+        'aria-labelledby': anchorId,
+        'aria-haspopup': 'listbox',
+        children: selectedItem?.text ?? placeholder,
+        onClick: onAnchorClick,
+        onKeyDown: onAnchorKeyDown
+      })}
+      {open && (
```

Comment:
I tend to prefer ternaries over `&&` for rendering JSX. Here's a good blog post explaining why: https://kentcdodds.com/blog/use-ternaries-rather-than-and-and-in-jsx.

Reply:
Related, from @smockle in https://github.com/primer/components/pull/1135#discussion_r614391663:
> I can accept [preferring ternaries over `&&`] as a general guideline. Ternaries are more explicit, and perhaps that will make PRC’s code more readable as a result, which is a good goal. Out of curiosity, [is anyone] aware of a lint rule we could use for this (would we want a lint rule?)

Reply:
Fantastic article with real world reasons why this is important! Fixed in https://github.com/primer/components/pull/1135/commits/710123afbb5548be9405e42d8465b80894764e43

Reply:
Fair enough. I want to propose a new operator, `!!`, which is the counterpart to `??`, basically "If LHS is null or undefined, evaluate to LHS, otherwise evaluate to RHS".